src/main/java/com/mixpanel/android/mpmetrics/MixpanelNotificationRouteActivity.java

```diff
@@ -0,0 +1,163 @@
+                    MPLog.d(LOGTAG, "Wanted to open url in browser but url is invalid: " + uri.toString() + ". Starting default intent");
+                    return defaultIntent;
+                }
+            case URL_IN_WEBVIEW:
+                if (isValidURL(uri.toString(), PushTapTarget.URL_IN_WEBVIEW)) {
+                    return new Intent(this.getApplicationContext(), MixpanelWebViewActivity.class).
+                            putExtra("uri", uri.toString()).
+                            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+                } else {
+                    MPLog.d(LOGTAG, "Wanted to open url in webview but url is invalid or not secure (http): " + uri.toString() + ". Starting default intent");
+                    return defaultIntent;
+                }
+            case DEEP_LINK:
+                return new Intent(Intent.ACTION_VIEW, Uri.parse(uri.toString()));
+            default:
+                return defaultIntent;
+        }
+    }
+
+    protected boolean isValidURL(CharSequence url, PushTapTarget target) {
+        if (target == PushTapTarget.URL_IN_WEBVIEW) {
+            return null != url && url.toString().startsWith("https");
+        } else {
+            return null != url && url.toString().startsWith("http");
+        }
+    }
+
+    protected void cancelNotification(Bundle extras) {
+        int notificationId = extras.getInt("notificationId");
+        NotificationManager notificationManager = (NotificationManager) getApplicationContext().getSystemService(getApplicationContext().NOTIFICATION_SERVICE);
+        notificationManager.cancel(notificationId);
+    }
+
+    protected void trackAction(Intent routeIntent) {
+        Bundle intentExtras = routeIntent.getExtras();
+
+        CharSequence tapTargetChars = intentExtras.getCharSequence("tapTarget");
+        if (null == tapTargetChars) {
+            MPLog.d(LOGTAG, "Notification action click logged with no tapTarget");
+            return;
+        }
+
+        final String tapTarget = tapTargetChars.toString();
+        final String buttonId;
+        final String label;
+
+        if (tapTarget.equals(MixpanelPushNotification.TAP_TARGET_BUTTON)) {
+            CharSequence buttonIdChars = intentExtras.getCharSequence("buttonId");
+            if (null == buttonIdChars) {
+                MPLog.d(LOGTAG, "Notification action click logged with no buttonId");
+                return;
+            }
+
+            CharSequence labelChars = intentExtras.getCharSequence("label");
+            if (null == labelChars) {
+                MPLog.d(LOGTAG, "Notification action click logged with no label");
+                return;
+            }
+
+            buttonId = buttonIdChars.toString();
+            label = labelChars.toString();
+        } else {
+            buttonId = null;
+            label = null;
+        }
+
+        CharSequence messageIdChars = intentExtras.getCharSequence("messageId");
+        if (null == messageIdChars) {
```

Comment:
same here, let's just always track and let nulls through. It _shouldn't_ ever happen but I think in the case where it were to happen we'd prefer not to lose the track call completely.